crates/shared/src/sources/uniswap_v2.rs

```diff
@@ -1,42 +1,102 @@
+        BaselineSource::UniswapV2 => {
+            Some((contracts::UniswapV2Router02::raw_contract(), UNISWAP_INIT))
+        }
+        BaselineSource::Honeyswap => {
+            Some((contracts::HoneyswapRouter::raw_contract(), HONEYSWAP_INIT))
+        }
+        BaselineSource::SushiSwap => {
+            Some((contracts::SushiSwapRouter::raw_contract(), SUSHISWAP_INIT))
+        }
+        BaselineSource::Baoswap => Some((contracts::BaoswapRouter::raw_contract(), BAOSWAP_INIT)),
+        BaselineSource::Swapr => Some((contracts::SwaprRouter::raw_contract(), SWAPR_INIT)),
+    }?;
+    let address = contract.networks.get(net_version)?.address;
+    Some((address, H256(init_code_digest)))
+}
 
-        // Görli
-        let (goerli_pair_provider, _) = get_liquidity_source(&Mock::new(5).web3()).await.unwrap();
-        let goerli_pair = TokenPair::new(
-            addr!("02ABBDbAaa7b1BB64B5c878f7ac17f8DDa169532"),
-            addr!("3430d04E42a722c5Ae52C5Bffbf1F230C2677600"),
-        )
-        .unwrap();
-        assert_eq!(
-            goerli_pair_provider.pair_address(&goerli_pair),
-            // https://goerli.etherscan.io/tx/0xd52899a351c83da758b944972b08f9fe1b856d723a9b2fae2a080fd83e29f386#eventlog
-            addr!("638F259D0A59e1d3b9e9f7E7dd1CB591C754005b")
-        );
+/// If the baseline source is uniswapv2-like, returns the corresponding
+/// PoolReading implementation.
+pub fn pool_reader(
+    source: BaselineSource,
+    pair_provider: PairProvider,
+    web3: &Web3,
+) -> Option<Box<dyn PoolReading>> {
+    let default_reader = DefaultPoolReader {
+        pair_provider,
+        web3: web3.clone(),
+    };
+    match source {
+        BaselineSource::BalancerV2 => None,
+        BaselineSource::ZeroEx => None,
+        BaselineSource::UniswapV3 => None,
+        BaselineSource::UniswapV2 => Some(Box::new(default_reader)),
+        BaselineSource::Honeyswap => Some(Box::new(default_reader)),
+        BaselineSource::SushiSwap => Some(Box::new(default_reader)),
+        BaselineSource::Baoswap => Some(Box::new(default_reader)),
+        BaselineSource::Swapr => Some(Box::new(SwaprPoolReader(default_reader))),
     }
 }
+
+pub async fn uniswap_like_liquidity_source(
+    source: BaselineSource,
+    web3: &Web3,
+) -> Result<Option<(PairProvider, Arc<dyn PoolFetching>)>> {
+    let net_version = web3.net().version().await.context("net_version")?;
+    let (router, init_code_digest) = match from_baseline_source(source, &net_version) {
+        Some(inner) => inner,
+        None => return Ok(None),
+    };
+    let router = contracts::IUniswapLikeRouter::at(web3, router);
+    let factory = router.factory().call().await.context("factory")?;
+    let provider = pair_provider::PairProvider {
+        factory,
+        init_code_digest: init_code_digest.0,
+    };
+    let pool_reader = match pool_reader(source, provider, web3) {
+        Some(inner) => inner,
+        None => return Ok(None),
```

Comment:
Wouldn't this be an error case? It probably doesn't make sense to be able to get the router but not the reader for a given liquidity source.
Also the implementation details should probably be encapsulated in this file so we could make most of the functions and constants private.

Reply:
Yes it would be an error. This stops mattering in the [follow up PR](https://github.com/cowprotocol/services/pull/1181) so I might not fix it here.